design.md

```diff
@@ -0,0 +1,73 @@
+# Bazel Module (bzlmod) Support in Renovate
+
+```yaml
+author: Chuck Grindel<chuck.grindel@gmail.com>
+status: To be reviewed
+created: 2023-04-10
+```
+
+## Abstract
+
+[Renovate] is a popular service for updating external dependencies. It [currently supports] updating
+Bazel dependencies described in [WORKSPACE-managed repositories]. With the introduction of [Bazel
+modules], [Renovate] needs to be modified to support the new external dependency mechanism.
+
+## Requirements
+
+- Support extracting external dependency information from `MODULE.bazel` files.
+  - Read external dependency information from [bazel_dep] declarations.
+  - Support module overrides.
+    - [Single version override]
+    - [Multiple version override]
+    - [Non-registry overrides]
+- Detect new module versions when they are introduced to a module registry (e.g., [Bazel Central
+  Registry]).
+  - Triggering updates based upon GitHub releases is problematic as the update to a registry is a
+    separate step that may not complete for minutes to hours (i.e., human approval of a pull
+    request).
+- Versioning
+  - Support [Bazel module version formats] and [compatibility level].
```

Comment:
Similar to above - do we also want to parse `.bazelversion` and apply checks against the `bazel_compatibility` attribute to the module definitions.

Reply:
I was going to call this out in the design section. However, I should call this out in the requirements.

So, the logic would be something like the following?
- If we find a `.bazelversion`, we will use the first value found in the file. (Off the top of my head, I believe that some folks have added other values to this file.) 
- If we have a Bazel version and `bazel_compatibility` is specified, we will perform the checks.
- No Bazel compatibility checks will occur if we do not have a Bazel version.

Reply:
The [doc for `bazel_compatibility`](https://bazel.build/rules/lib/globals#module) states:

> A list of bazel versions that allows users to declare which Bazel versions are compatible with this module. It does NOT affect dependency resolution, but bzlmod will use this information to check if your current Bazel version is compatible. 

The "does NOT affect dependency resolution" is interesting.

@Wyverald @meteorcloudy Does this mean that bzlmod could resolve to a module version that is not compatible with the current Bazel version? If one directly specifies a version for a module that is not compatible, will Bazel fail on startup?

It seems like it would be reasonable for the Renovate manager not to suggest a Bazel version that is incompatible. Can anyone think of a reason not to use `bazel_compatibility` as a way to filter out upgrade candidates?

Reply:
I don't think we can filter them out, at least by default, otherwise nothing will ever be upgraded (a circle of compatibility restrictions - x depends on y+1, y depends on x+1 etc).

I think the general philosophy is to detect and propose version upgrades regardless of compatibility, and the user adds restrictions on top of that if necessary. For example https://docs.renovatebot.com/configuration-options/#constraints

Reply:
@jbedard Are you suggesting someone could author constraints preventing a module from upgrading? If so, I agree. However, I am not sure that means we should not respect their constraints.

Also, from the doc that you linked, I noticed the following:

> Constraints are used in package managers which use third-party tools to update "artifacts" like lock files or checksum files. Typically, the constraint is detected automatically by Renovate from files within the repository and there is no need to manually configure it.

I think that respecting the `bazel_compatibility` constraints falls into the category of "automatically detected".

To make this more concrete, let's look at an example. Let's assume that a repository uses Bazel 6.0.0 and has the following dependency.

```python
bazel_dep(name = "rules_foo", version = "1.0.0")
```

Now, the maintainers for `rules_foo` release a new version, 1.1.0, with the following module definition.

```python
module(
    name = "rules_foo",
    version = "1.1.0",
    bazel_compatibility = [">=6.1.0"],
)
```

The question is: Should Renovate propose upgrading `rules_foo` to 1.1.0?

If we respect the `bazel_compatibility`, then Renovate would not propose this upgrade until the client repository started using Bazel 6.1.0 or above.

If we ignore the `bazel_compatibility`, then Renovate will propose the upgrade and Bazel will fail due to the incompatible Bazel version. (I think.)

Reply:
> Typically, the constraint is detected automatically by Renovate from files within the repository

I think that is to respect constraints of the repo being updated, not constraints of the dependencies.

If respecting the constraints of dependencies that means we must fetch the entire dep graph and find a version compatible with everything? I think we should only need to know the dependencies declared in the repo and the available versions of those dependencies.

Reply:
Sorry for being late to the party, but:

> Does this mean that bzlmod could resolve to a module version that is not compatible with the current Bazel version? If one directly specifies a version for a module that is not compatible, will Bazel fail on startup?

Yes. The `bazel_compatibility` field is not taken into account during module version resolution, and is performed as a post-resolution check instead.